solr/core/src/java/org/apache/solr/managed/ResourceManagerPool.java

```diff
@@ -0,0 +1,216 @@
+  protected final Class<? extends ManagedComponent> componentClass;
+  private final Map<String, Object> poolParams;
+  protected final ResourcePoolContext poolContext = new ResourcePoolContext();
+  protected final List<ChangeListener> listeners = new ArrayList<>();
+  protected final ReentrantLock updateLock = new ReentrantLock();
+  protected int scheduleDelaySeconds;
+  protected ScheduledFuture<?> scheduledFuture;
+
+  public ResourceManagerPool(String name, String type, ResourceManager resourceManager,
+                                Map<String, Object> poolLimits, Map<String, Object> poolParams) {
+    this.name = name;
+    this.type = type;
+    this.resourceManager = resourceManager;
+    this.componentClass = resourceManager.getResourceManagerPoolFactory().getComponentClassByType(type);
+    this.poolLimits = new HashMap<>(poolLimits);
+    this.poolParams = new HashMap<>(poolParams);
+  }
+
+  /** Unique pool name. */
+  public String getName() {
+    return name;
+  }
+
+  /** Pool type. */
+  public String getType() {
+    return type;
+  }
+
+  public ResourceManager getResourceManager() {
+    return resourceManager;
+  }
+
+  /** Add component to this pool. */
+  public void registerComponent(T managedComponent) {
+    if (!componentClass.isAssignableFrom(managedComponent.getClass())) {
+      log.debug("Pool type '" + type + "' is not supported by the component " + managedComponent.getManagedComponentId());
+      return;
+    }
+    ManagedComponent existing = components.putIfAbsent(managedComponent.getManagedComponentId().toString(), managedComponent);
+    if (existing != null) {
+      throw new IllegalArgumentException("Component '" + managedComponent.getManagedComponentId() + "' already exists in pool '" + name + "' !");
+    }
+  }
+
+  /** Remove named component from this pool. */
+  public boolean unregisterComponent(String componentId) {
+    return components.remove(name) != null;
+  }
+
+  /**
+   * Check whether a named component is registered in this pool.
+   * @param componentId component id
+   * @return true if the component with this name is registered, false otherwise.
+   */
+  public boolean isRegistered(String componentId) {
+    return components.containsKey(componentId);
+  }
+
+  /** Get components managed by this pool. */
+  public Map<String, T> getComponents() {
+    return Collections.unmodifiableMap(components);
+  }
+
+  public void addChangeListener(ChangeListener listener) {
+    if (!listeners.contains(listener)) {
+      listeners.add(listener);
+    }
+  }
+
+  public void removeChangeListener(ChangeListener listener) {
+    listeners.remove(listener);
+  }
+
+
+  /**
+   * Get the current monitored values from all resources. Result is a map with resource names as keys,
+   * and param/value maps as values.
+   */
+  public Map<String, Map<String, Object>> getCurrentValues() throws InterruptedException {
+    updateLock.lockInterruptibly();
```

Comment:
I'm having trouble understanding why the updateLock is necessary in this method? It is iterating over the values of a concurrent hash map (`components`) and reading the id and monitored value of the component. Both seem to be thread-safe to me.

Reply:
The updateLock is acquired only in getCurrentValues() (called by doManage and ResourceManagerHandler.STATUS) and in manage() but manage() is called only by a scheduled executor which guarantees a happens-before relationship to previous executions. I think we can get rid of the lock entirely?

Reply:
Yes, it's a left-over from a previous refactoring.